### PR TITLE
[BD-21] Refactor annotation checking to expose errors to pylint

### PR DESCRIPTION
**Pull request by regisb**
**Description:** Instead of just storing the error messages as part of `check_results()`, we
also store the type of error. This allows us to expose these errors to pylint
in edx-lint, with the help of a new checker. (to be created)

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation

**Dependencies:** a corresponding edx-lint PR will be created, which will depend on this one: https://github.com/edx/edx-lint/pull/140

**Reviewers:**
- [ ] @robrap 
- [ ] @bmedx 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-745](https://openedx.atlassian.net/browse/BLENDED-745) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


When this pull request is ready, tag your edX technical lead.

File: code_annotations/base.py [hunk 1]
```diff
@@ -314,7 +315,11 @@ def __init__(self, config):
         """
         self.config = config
         self.echo = self.config.echo
+        # errors contains formatted error messages
         self.errors = []
+        # annotation_errors contains (annotation, AnnotationErrorType, args) tuples
+        # This attribute may be parsed by 3rd-parties, such as edx-lint.
+        self.annotation_errors = []
```
Comment (robrap) [inline code_annotations/base.py:322]: If this is required to keep edx-lint happy, is it possible to unit test to ensure it keeps working?

**Review by robrap — COMMENTED** [review]
Great.  Thanks so much.  This will be a big win.

**Comment by robrap** [comment]
FYI: @bmedx: I'm not sure where you may have added code-annotations linting in the past, and whether our bringing this into edx-lint will enable you to clean-up/remove potentially redundant linting anywhere.

File: code_annotations/annotation_errors.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    error_type = AnnotationErrorType(
+        message,
+        symbol,
+        description,
+    )
+    TYPES.append(error_type)
+    return error_type
+
+
+# Choice errors
```
[the remark this answers is not in the document]
Comment (regisb) [reply code_annotations/annotation_errors.py:28]: I can certainly add an explicit comment here, but I don't really know what such a unit test would look like.

File: code_annotations/base.py (hunk 1 above)
Comment (regisb) [reply code_annotations/base.py:322]: :ok_hand:

**Comment by regisb** [comment]
@robrap See the updated PR. I squashed my commits to hasten merging.

> UPDATE: @regisb: Maybe what I wrote isn't true. Does this only apply to annotation config files that live under contrib and are added to edx-lint?

Generic code annotation error checking can be performed for all configuration files, but pylinting will only work with the configuration files from contrib/config.
I believe that it's still a good idea to run `code_annotations --lint` whenever we extract code annotations, just to make sure that they are correctly formatted at runtime.

File: code_annotations/base.py [hunk 3]
```diff
@@ -524,69 +529,26 @@ def check_group(self, annotations):
                 if token not in found_tokens:
                     self._add_annotation_error(
                         annotations[0],
-                        "missing non-optional annotation: '{}'".format(token)
-                    )
-
-    def check_annotation(self, annotation, current_group):
```
Comment (bmedx) [inline code_annotations/base.py:530]: Wow looks like this was just completely unused, eh?

**Review by bmedx — APPROVED** [review]
Looks good to me, much cleaner and the edx-lint integration should be super helpful. 👍🏼

File: code_annotations/base.py (hunk 3 above)
Comment (regisb) [reply code_annotations/base.py:530]: Yes, I had forgotten to remove it after my previous cleanup.

**Review by robrap — APPROVED** [review]
The unit test idea is not a blocker. Feel free to merge.

A unit test could simply declare an expected list of types that is compared to TYPES. It would break on any new item until it is added to the end of expected, but a test failure may call more attention to the importance of the order (the test name and docstring could both note this), than the code comment that might not be read if someone adds to the middle of the list.

**Comment by regisb** [comment]
> The unit test idea is not a blocker. Feel free to merge.

I just realized that I did not push the new test module to my branch... Updating this PR now.

**Pull request merged**